### PR TITLE
Activations - relu6 & hardTanh

### PR DESCRIPTION
**Pull request by asilvas**

### CODE REVIEW

File: shumai/tensor/register_gradients.ts [hunk 1]
```diff
@@ -210,6 +210,18 @@ const impls = {
     const o = sm.scalar(1).sub(grad.out)
     return grad.out.mul(o)
   },
+  clip: (grad: Grad) => {
```
Comment (asilvas) [inline shumai/tensor/register_gradients.ts:213]: Still getting the hang of how gradients are implemented in Shumai, so not certain this is correct. Probably wise to eventually have tests around these gradient functions.

File: shumai/tensor/tensor_ops.ts [hunk 2]
```diff
@@ -36,3 +36,15 @@ export function thresholdRelu(tensor: Tensor, threshold = 1): Tensor {
   const mask = tensor.gte(scalar(threshold))
   return tensor.mul(mask)
 }
+
+export function clamp(tensor: Tensor, low: number, high: number): Tensor {
+  return tensor.clip(full(tensor.shape, low), full(tensor.shape, high));
```
Comment (asilvas) [inline shumai/tensor/tensor_ops.ts:41]: I wasn't aware `clip` did not work with scalars -- not sure why that is. `clamp` might be a little redundant but kept for consistency with FL.

File: shumai/tensor/register_gradients.ts [hunk 3]
```diff
@@ -210,6 +210,18 @@ const impls = {
...
+    const result = <Tensor>grad.in[0]
+    const low = <Tensor>grad.in[1]
+    const high = <Tensor>grad.in[2]
+
+    const lowMask = result.greaterThan(low);
+    const highMask = result.lessThan(high);
+    const lowHighMask = lowMask.bitwiseAnd(highMask);
+    const gradMask = sm.where(lowHighMask, grad.out, sm.full(grad.out.shape, 0))
```
Comment (bwasti) [inline shumai/tensor/register_gradients.ts:221]: looks right to me

**Review by bwasti — APPROVED** [review]

**Comment by cryptodeal** [comment]
@asilvas, if you've got a Discord, you should join the [Shumai dev Server](https://discord.gg/ZYxe8dQE); we might be able to collaborate on getting a Shumai implementation of autograd checks implemented (essentially Shumai equivalent of: [torch.autograd.gradcheck](https://pytorch.org/docs/stable/generated/torch.autograd.gradcheck.html)

**Pull request merged**